echopype/utils/io.py

```diff
@@ -414,24 +417,46 @@ def validate_source_ds_da(
 
     # check that source_ds_da is of the correct type, if it is a path validate
     # the path and open the Dataset or DataArray using xarray
-    if not isinstance(source_ds_da, (xr.Dataset, xr.DataArray, str, Path)):
-        raise TypeError("source_ds_da must be a Dataset or DataArray or str or pathlib.Path!")
-    elif isinstance(source_ds_da, (str, Path)):
+    if not isinstance(source_ds_da_ed, (xr.Dataset, xr.DataArray, EchoData, str, Path)):
+        raise TypeError(
+            "source_ds_da must be a Dataset or DataArray or EchoData or str or pathlib.Path!"
+        )
+    elif isinstance(source_ds_da_ed, (str, Path)):
         # determine if we obtained a zarr or netcdf file
-        file_type = get_file_format(source_ds_da)
+        file_type = get_file_format(source_ds_da_ed)
 
         # validate source_ds_da if it is a path
-        source_ds_da = validate_output_path(
+        source_ds_da_ed = validate_output_path(
             source_file="blank",  # will be unused since source_ds cannot be none
             engine=file_type,
             output_storage_options=storage_options,
-            save_path=source_ds_da,
+            save_path=source_ds_da_ed,
         )
 
         # check that the path exists
-        check_file_existence(file_path=source_ds_da, storage_options=storage_options)
+        check_file_existence(file_path=source_ds_da_ed, storage_options=storage_options)
+
+    return source_ds_da_ed, file_type
+
+
+def open_ds_da_ed_path(
+    source_ds_da: Optional[Union[xr.Dataset, xr.DataArray, str, Path]],
+    source_ed: Optional[Union[EchoData, str, Path]],
+    storage_options: Optional[dict],
+) -> Tuple[Optional[Union[xr.Dataset, xr.DataArray]], Optional[EchoData]]:
```

Comment:
Somehow I feel forcing `source_ed` to always be in the input argument is a little awkward, when you only need to deal with `source_ds_da`. Do you think it will work if the function is changed to take in `source`, `source_type`, and `storage_options`, and depending on `source_type` different open function is used.

Reply:
> Do you think it will work if the function is changed to take in `source`, `source_type`, and `storage_options`, and depending on `source_type` different open function is used.

@leewujung Yes, I have updated the method which will use different open function depending on `source_type`.